src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.ReadCore.cs

```diff
@@ -14,36 +14,17 @@ public partial class JsonConverter<T>
             {
                 if (!state.IsContinuation)
                 {
-                    if (!SingleValueReadWithReadAhead(RequiresReadAhead, ref reader, ref state))
+                    // This is first call to the converter -- advance the reader
+                    // to the first JSON token and perform a read-ahead if necessary.
+                    if (!reader.TryAdvanceWithOptionalReadAhead(RequiresReadAhead))
                     {
-                        if (state.SupportContinuation)
+                        if (state.SupportContinuation && state.Current.ReturnValue is object result)
                         {
-                            // If a Stream-based scenario, return the actual value previously found;
-                            // this may or may not be the final pass through here.
-                            state.BytesConsumed += reader.BytesConsumed;
-                            if (state.Current.ReturnValue == null)
-                            {
-                                // Avoid returning null for value types.
-                                return default;
-                            }
-
-                            return (T)state.Current.ReturnValue!;
-                        }
-                        else
-                        {
-                            // Read more data until we have the full element.
-                            state.BytesConsumed += reader.BytesConsumed;
-                            return default;
+                            // This branch is hit when deserialization has completed in an earlier call
+                            // but we're still processing trailing whitespace. Return the result stored in the state machine.
+                            return (T)result;
                         }
-                    }
-                }
-                else
-                {
-                    // For a continuation, read ahead here to avoid having to build and then tear
-                    // down the call stack if there is more than one buffer fetch necessary.
-                    if (!SingleValueReadWithReadAhead(requiresReadAhead: true, ref reader, ref state))
```

Comment:
Here's the crux of the second change made by this PR: re-entrant serialization operations no longer perform read-ahead of the next value (the converter doesn't require it in this case). We let the `JsonSerializerOptions.DefaultBufferSize` be the only guide of how much data we add to the buffer before re-entering deserialization.